api/tokens.go

```diff
@@ -165,3 +219,74 @@ func handleUnexpectedError(ctx context.Context, w http.ResponseWriter, err error
 	writeErrorResponse(ctx, w, http.StatusInternalServerError, errorResponseBody)
 	return
 }
+
+func computeSecretHash(clientSecret string, username string, clientId string) string {
+	mac := hmac.New(sha256.New, []byte(clientSecret))
+	mac.Write([]byte(username + clientId))
+
+	return base64.StdEncoding.EncodeToString(mac.Sum(nil))
+}
+
+func buildCognitoRequest(authParams AuthParams, clientId string, clientSecret string, clientAuthFlow string) (authInput *cognitoidentityprovider.InitiateAuthInput) {
+
+	secretHash := computeSecretHash(clientSecret, authParams.Email, clientId)
+
+	authParameters := map[string]*string{
+		"USERNAME":    &authParams.Email,
+		"PASSWORD":    &authParams.Password,
+		"SECRET_HASH": &secretHash,
+	}
+
+	authInput = &cognitoidentityprovider.InitiateAuthInput{
+		AnalyticsMetadata: &cognitoidentityprovider.AnalyticsMetadataType{},
+		AuthFlow:          &clientAuthFlow,
+		AuthParameters:    authParameters,
+		ClientId:          &clientId,
+		ClientMetadata:    map[string]*string{},
+		UserContextData:   &cognitoidentityprovider.UserContextDataType{},
+	}
+
+	return authInput
+}
+
+func buildSucessfulResponse(result *cognitoidentityprovider.InitiateAuthOutput, w http.ResponseWriter, ctx context.Context) {
+
+	if result.AuthenticationResult != nil {
```

Comment:
If this condition is not met the endpoint will return an empty 200 response.
Probably a good idea to have an else to catch anything unexpected from Cognito and return a 500

Reply:
Yes, well spotted. Thank you for the suggestion